File: src/main/java/managers/CommodityManager.java
package managers;

import cashshop.commodities.Commodity;
import cashshop.commodities.CommodityFlag;
import util.packet.PacketReader;

import java.util.HashMap;
import java.util.Map;

public class CommodityManager extends AbstractManager {

    private static final Map<Integer, Commodity> commodities = new HashMap<>();

    public static synchronized Commodity getCommodity(int sn) {
        Commodity commodity = commodities.get(sn);
        if (commodity == null) {
            commodity = new Commodity(sn);
            if (!loadData(commodity)) {
                return null;
            }
            commodities.put(sn, commodity);
        }
        return commodity;
    }

    private static boolean loadData(Commodity commodity) {
        PacketReader r = getData("wz/Commodity/" + commodity.getSN() + ".mao");
        if (r != null) {
            int flags = r.readInteger();
            r.readInteger(); // SN
            commodity.setItemId(r.readInteger());
            if (containsFlag(flags, CommodityFlag.COUNT)) commodity.setCount(r.readShort());
            if (containsFlag(flags, CommodityFlag.PRIORITY)) commodity.setPriority(r.readByte());
            if (containsFlag(flags, CommodityFlag.PRICE)) commodity.setPrice(r.readInteger());
            if (containsFlag(flags, CommodityFlag.BONUS)) commodity.setBonus(r.readByte());
            if (containsFlag(flags, CommodityFlag.PERIOD)) commodity.setPeriod(r.readShort());
            if (containsFlag(flags, CommodityFlag.REQ_POP)) commodity.setReqPop(r.readShort());
            if (containsFlag(flags, CommodityFlag.REQ_LEV)) commodity.setReqLev(r.readShort());
            if (containsFlag(flags, CommodityFlag.MAPLE_POINT)) commodity.setMaplePoint(r.readInteger());
            if (containsFlag(flags, CommodityFlag.MESO)) commodity.setMeso(r.readInteger());
            if (containsFlag(flags, CommodityFlag.FOR_PREMIUM_USER)) commodity.setForPremiumUser(r.readBool());
            if (containsFlag(flags, CommodityFlag.GENDER)) commodity.setGender(r.readByte());
            if (containsFlag(flags, CommodityFlag.ON_SALE)) commodity.setOnSale(r.readBool());
            if (containsFlag(flags, CommodityFlag.CLASS)) commodity.setJob(r.readByte());
            if (containsFlag(flags, CommodityFlag.LIMIT)) commodity.setLimit(r.readByte());
            if (containsFlag(flags, CommodityFlag.PB_CASH)) commodity.setPbCash(r.readShort());
            if (containsFlag(flags, CommodityFlag.PB_POINT)) commodity.setPbPoint(r.readShort());
            if (containsFlag(flags, CommodityFlag.PB_GIFT)) commodity.setPbGift(r.readShort());

            return true;
        } else {
            System.err.println("Commodity " + commodity.getSN() + " does not exist!");
        }
        return false;
    }

    private static boolean containsFlag(int flags, CommodityFlag flag) {
        return (flags & flag.getValue()) == flag.getValue();
    }
}

File: src/main/java/managers/CommodityManager.kt
package managers

import cashshop.commodities.Commodity
import cashshop.commodities.CommodityFlag

object CommodityManager : AbstractManager() {

    private val commodities: MutableMap<Int, Commodity> = HashMap()

    fun getCommodity(sn: Int): Commodity? {
        synchronized(commodities) {
            var commodity = commodities[sn]

            if (commodity == null) {
                commodity = Commodity(sn)
                if (!loadData(commodity)) {
                    System.err.println("Commodity " + commodity.SN + " does not exist!")
                    return null
                }
                commodities[sn] = commodity
            }

            return commodity
        }
    }

    private fun loadData(commodity: Commodity): Boolean {
        val r = getData("wz/Commodity/" + commodity.SN + ".mao") ?: return false

        val flags = r.readInteger()
        r.readInteger() // SN
        commodity.itemId = r.readInteger()
        if (containsFlag(flags, CommodityFlag.COUNT)) commodity.count = r.readShort()
        if (containsFlag(flags, CommodityFlag.PRIORITY)) commodity.priority = r.readByte()
        if (containsFlag(flags, CommodityFlag.PRICE)) commodity.price = r.readInteger()
        if (containsFlag(flags, CommodityFlag.BONUS)) commodity.bonus = r.readByte()
        if (containsFlag(flags, CommodityFlag.PERIOD)) commodity.period = r.readShort()
        if (containsFlag(flags, CommodityFlag.REQ_POP)) commodity.reqPop = r.readShort()
        if (containsFlag(flags, CommodityFlag.REQ_LEV)) commodity.reqLev = r.readShort()
        if (containsFlag(flags, CommodityFlag.MAPLE_POINT)) commodity.maplePoint = r.readInteger()
        if (containsFlag(flags, CommodityFlag.MESO)) commodity.meso = r.readInteger()
        if (containsFlag(flags, CommodityFlag.FOR_PREMIUM_USER)) commodity.isForPremiumUser = r.readBool()
        if (containsFlag(flags, CommodityFlag.GENDER)) commodity.gender = r.readByte()
        if (containsFlag(flags, CommodityFlag.ON_SALE)) commodity.isOnSale = r.readBool()
        if (containsFlag(flags, CommodityFlag.CLASS)) commodity.job = r.readByte()
        if (containsFlag(flags, CommodityFlag.LIMIT)) commodity.limit = r.readByte()
        if (containsFlag(flags, CommodityFlag.PB_CASH)) commodity.pbCash = r.readShort()
        if (containsFlag(flags, CommodityFlag.PB_POINT)) commodity.pbPoint = r.readShort()
        if (containsFlag(flags, CommodityFlag.PB_GIFT)) commodity.pbGift = r.readShort()

        return true

    }

    private fun containsFlag(flags: Int, flag: CommodityFlag): Boolean {
        return flags and flag.value == flag.value
    }
}